packages/bus/src/Brokers/RMQEventBus.ts

```diff
@@ -0,0 +1,395 @@
+import {
+  Command,
+  Event,
+  EventBus,
+  EventHandler,
+  NoResponse,
+  RetryStrategy,
+  Configuration,
+  Credentials,
+  IllegalOperation,
+  EventHandlerNotFound,
+  NoSubscriptionsFound,
+  EventHandlerConstructor
+} from '@secbox/core';
+import { ConfirmChannel, Connection, ConsumeMessage } from 'amqplib';
+import { DependencyContainer, inject, injectable } from 'tsyringe';
+import { EventEmitter, once } from 'events';
+import { format, parse, UrlWithParsedQuery } from 'url';
+
+interface ParsedConsumeMessage<T = unknown> {
+  payload: Event<T>;
+  name: string;
+  replyTo?: string;
+  correlationId?: string;
+}
+
+@injectable()
+export class RMQEventBus implements EventBus {
+  private client?: Connection;
+  private channel?: ConfirmChannel;
+
+  private readonly subject = new EventEmitter({ captureRejections: true });
+  private readonly handlers = new Map<
+    string,
+    EventHandler<unknown, unknown>[]
+  >();
+  private readonly consumerTags: string[] = [];
+  private readonly container!: DependencyContainer;
+  private readonly backOffRetry!: RetryStrategy;
+
+  private readonly REPLY_QUEUE_NAME = 'amq.rabbitmq.reply-to';
+  private readonly APP_QUEUE_NAME = 'app';
+  private readonly DEFAULT_RECONNECT_TIMES = 20;
+
+  constructor(
+    @inject(Configuration) private readonly configuration: Configuration,
+    @inject(ConnectionFactory)
+    private readonly connectionFactory: ConnectionFactory<Connection>
+  ) {
+    this.container = configuration.container;
+    this.subject.setMaxListeners(Infinity);
+    this.backOffRetry = new ExponentialBackoffRetryStrategy({
+      maxDepth: this.DEFAULT_RECONNECT_TIMES
+    });
+  }
+
+  public async init(): Promise<void> {
+    const connectTimeout = this.configuration.connectTimeout;
+    const heartbeat = this.configuration.heartbeatInterval;
+    const credentials = this.configuration.credentials;
+    const bus = this.configuration.bus;
+
+    const buildedUrl = this.buildUrl(bus, credentials, heartbeat || 5000);
+
+    this.client = await this.backOffRetry.acquire(async () =>
+      this.connectionFactory.create(buildedUrl, {
+        timeout: connectTimeout
+      })
+    );
+
+    await this.createConsumerChannel();
+  }
+
+  public async register<T, R>(
+    type: EventHandlerConstructor<T, R>
+  ): Promise<void> {
+    if (!this.channel) {
+      throw new IllegalOperation(this);
+    }
+
+    if (!this.container.isRegistered(type)) {
+      throw new EventHandlerNotFound(type.name);
+    }
+    const handlerInstance = this.container.resolve(type);
+
+    const eventNames: string[] = this.reflectEventsNames(type);
+    if (eventNames?.length === 0) {
+      throw new NoSubscriptionsFound(handlerInstance);
+    }
+
+    for (const eventName of eventNames) {
+      const eventHandlers = this.handlers.get(eventName);
+      if (Array.isArray(eventHandlers)) {
+        eventHandlers.push(handlerInstance);
+      } else {
+        await this.channel?.bindQueue(
```

Comment:
You have already asserted channel a few lines above

```suggestion
        await this.channel.bindQueue(
```

Reply:
The channel is declared as optional, and if we delete this assertion check it means that we need to add the assertion check above in `if` scope

Reply:
Yeah, that's why you already have the corresponding `if` block on the top of the method)

Reply:
Yes, but for some reason build failed in the case when I remove it assertion check. It wants corresponding `if` block on the same nested level

Reply:
Well, it has nothing in common with nesting levels or blocks. The compiler raises an exception if you use such a variable inside the async function since it might be null because of the nature of the async code.

Good:
```
let a: string | undefined;

if (!a) {
  throw new Error()
}

for (;;) {
  a.toLowerCase();
}
```

Bad:

```
let a: string | undefined;

if (!a) {
  throw new Error()
}

process.nextTick(() => a.toLowerCase());
```